### PR TITLE
Return 410 error when data store that has been deleted is requested externally

### LINKED ISSUE #4824: Do not return data that has been flagged as unreferenced by the GC process
## Work Item
**Describe the outcome you expect**
If a client makes a request for a data store/component and that object has been marked as unreferenced by the GC process, the request should fail.

**Approach**
Option 1: ask SPO to return the unreferenced flag and use that information to fail requests for deleted content
Option 2: manage the referenced flags generated by the GC process at summarizer time in the runtime and use that info to fail the requests.

**Acceptance criteria**
Requests for deleted content will result in a 404.
The deleted content is never sent to the browser.

### PR DESCRIPTION
**Pull request by agarwal-navin**
Fixes #4824.

Return a 410 (Gone) error code when a data store is requested that has been deleted (marked unreferenced after GC).
We do this only for external requests, i.e., requests that come from Container (loader layer). Container adds a header into the request that says return error if the content being requested is deleted.
Data store runtime returns 410 if the above header is set and the data store is unreferenced as per initial summary.

P.S. I am working on adding end-to-end tests for this.

### CODE REVIEW

File: packages/loader/container-loader/src/container.ts [hunk 1]
```diff
@@ -158,6 +158,23 @@ export enum ConnectionState {
     Connected,
 }
 
+// Gate that when flipped, instructs to return an error if the resource being requested has been deleted.
+function gatesErrorOnDeletedResourceRequest() {
```
Comment (agarwal-navin) [inline packages/loader/container-loader/src/container.ts:162]: We should probably have a generic function that can do this.

File: packages/loader/container-loader/src/container.ts [hunk 2]
```diff
@@ -776,9 +793,23 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
...
-    public async request(path: IRequest): Promise<IResponse> {
+    public async request(request: IRequest): Promise<IResponse> {
+        /**
+         * For external requests coming from Loader layer, add a header that will enable returning an error if the
+         * resource being requested has been deleted because it is unreferenced.
+         *
+         * This is used to handle scenario where a resource shared with an external app is deleted and hence GC'd.
+         * The error returned signals the app to take appropriate action for deleted content.
+         */
+        if (gatesErrorOnDeletedResourceRequest()) {
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/container.ts:804]: why do we need this. can't bohemia just provide the header?

File: packages/runtime/runtime-definitions/src/dataStoreContext.ts [hunk 3]
```diff
@@ -347,6 +347,11 @@ IEventProvider<IFluidDataStoreContextEvents>, Partial<IProvideFluidDataStoreRegi
      * and its children with the GC details from the previous summary.
      */
     getInitialGCSummaryDetails(): Promise<IGarbageCollectionSummaryDetails>;
+
+    /**
+     * Tells whether the data store is referenced as per the initial summary when it was loaded.
+     */
+    isReferencedInInitialSummary(): Promise<boolean>;
```
Comment (anthony-murphy) [inline packages/runtime/runtime-definitions/src/dataStoreContext.ts:354]: do we need this on the interface? can we get by with this just on the implementation?

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 4]
```diff
@@ -346,6 +346,22 @@ IFluidDataStoreChannel, IFluidDataStoreRuntime, IFluidHandleContext {
...
         }
 
+        /**
+         * If the request has "errorOnResourceDeleted" header, we need to return an error if the data store being
+         * requested has been deleted.
+         *
+         * This is used to handle scenario where a data store shared with an external app is deleted and hence GC'd.
+         * The error returned signals the app to take appropriate action for deleted content.
+         */
+        if (request.headers?.errorOnResourceDeleted === true) {
```
Comment (anthony-murphy) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:356]: let's be more explicit with the header name, maybe something like errorOnUnreferencedResources

File: packages/runtime/runtime-definitions/src/dataStoreContext.ts (hunk 3 above)
Comment (anthony-murphy) [reply packages/runtime/runtime-definitions/src/dataStoreContext.ts:354]: i really don't think this is the right design, so would like to minimize this as much as possible

**Pull request closed without merging**